sound/soc/sof/ipc4-topology.c

```diff
@@ -1387,6 +1469,86 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct sof_ipc4_process *process = swidget->private;
+	struct sof_ipc4_available_audio_format *available_fmt = &process->available_fmt;
+	struct sof_ipc4_control_data *control_data;
+	struct snd_sof_control *scontrol = NULL;
+	void *cfg = process->ipc_config_data;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	void *data;
+	int ret, i;
+
+	available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules setup the output format based on audio format tokens in topology.
+	 */
+	if (swidget->payload_with_output)
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      &process->output_format, pipeline_params,
+					      available_fmt,
+					      sizeof(struct sof_ipc4_base_module_cfg));
+	else
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      NULL, pipeline_params, available_fmt,
+					      sizeof(struct sof_ipc4_base_module_cfg));
+	if (ret < 0)
+		return ret;
+
+	/* update pipeline memory usage */
+	sof_ipc4_update_pipeline_mem_usage(sdev, swidget, &process->base_config);
+
+	/*
+	 * ipc_config_data is composed of the base_config, optional output formats followed
+	 * by the data required for module init in that order.
+	 */
+	memcpy(cfg, &process->base_config, sizeof(struct sof_ipc4_base_module_cfg));
+	cfg += sizeof(struct sof_ipc4_base_module_cfg);
+
+	/* copy output format to configure data payload */
+	if (swidget->payload_with_output) {
+		memcpy(cfg, &process->output_format, sizeof(struct sof_ipc4_audio_format));
+		cfg += sizeof(struct sof_ipc4_audio_format);
+	}
+
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+		if (!kc)
+			return -ENODEV;
```

Comment:
when would this be the case? And if so is it an error? can we skip and move on?

Reply:
Let me remove it. Suppose this should never happen.